modules/dnn/src/nms.cpp

```diff
@@ -58,6 +58,73 @@ void NMSBoxes(const std::vector<RotatedRect>& bboxes, const std::vector<float>&
     NMSFast_(bboxes, scores, score_threshold, nms_threshold, eta, top_k, indices, rotatedRectIOU);
 }
 
+template<typename T>
+static inline float getMaxCoord(const std::vector<Rect_<T>>& bboxes)
+{
+    double x1, y1, x2, y2, max_coord = 0;
+    for (int i = 0; i < bboxes.size(); i++)
+    {
+        x1 = bboxes[i].x;
+        y1 = bboxes[i].y;
+        x2 = x1 + bboxes[i].width;
+        y2 = y1 + bboxes[i].height;
+
+        max_coord = std::max(x1, max_coord);
+        max_coord = std::max(y1, max_coord);
+        max_coord = std::max(x2, max_coord);
+        max_coord = std::max(y2, max_coord);
+    }
+    return max_coord;
+}
+
+void NMSBoxesBatched(const std::vector<Rect>& bboxes, const std::vector<float>& scores, const std::vector<int>& class_ids,
+                     const float score_threshold, const float nms_threshold,
+                     std::vector<int>& indices, const float eta, const int top_k)
+{
+    CV_Assert_N(bboxes.size() == scores.size(), scores.size() == class_ids.size(), nms_threshold >= 0, eta > 0);
+
+    // find the max coordinate
+    int max_coord = getMaxCoord<int>(bboxes);
+
+    // calculate offset and add offset to each bbox
+    std::vector<Rect2i> bboxes_offset;
+    int offset;
+    for (int i = 0; i < bboxes.size(); i++)
+    {
+        offset = class_ids[i] * (max_coord + 1);
+        bboxes_offset.push_back(
+            Rect2i(bboxes[i].x + offset, bboxes[i].y + offset,
+                   bboxes[i].width, bboxes[i].height)
+        );
+    }
+
+    NMSFast_(bboxes_offset, scores, score_threshold, nms_threshold, eta, top_k, indices, rectOverlap);
+}
+
+void NMSBoxesBatched(const std::vector<Rect2d>& bboxes, const std::vector<float>& scores, const std::vector<int>& class_ids,
+                     const float score_threshold, const float nms_threshold,
+                     std::vector<int>& indices, const float eta, const int top_k)
+{
+    CV_Assert_N(bboxes.size() == scores.size(), scores.size() == class_ids.size(), nms_threshold >= 0, eta > 0);
+
+    // find the max coordinate
+    float max_coord = getMaxCoord<double>(bboxes);
+
+    // calculate offset and add offset to each bbox
+    std::vector<Rect2d> bboxes_offset;
+    float offset;
+    for (int i = 0; i < bboxes.size(); i++)
+    {
+        offset = class_ids[i] * (max_coord + 1);
+        bboxes_offset.push_back(
+            Rect2d(bboxes[i].x + offset, bboxes[i].y + offset,
+                   bboxes[i].width, bboxes[i].height)
+        );
+    }
+
+    NMSFast_(bboxes_offset, scores, score_threshold, nms_threshold, eta, top_k, indices, rectOverlap);
+}
```

Comment:
I would suggest to make a new template function instead of `getMaxCoord`, that has the signature

```cpp

template <class T>
void NMSBoxesBatched(const std::vector<T>& bboxes, const std::vector<float>& scores, const std::vector<int>& class_ids,
                     const float score_threshold, const float nms_threshold,
                     std::vector<int>& indices, const float eta, const int top_k)
                     
```

And just call it twice. There is `Rect_<T>::value_type` to get the coord type. Also, why are `max_coord` and `offset` floats in case of Rect_<double>?

Reply:
Made a template called `NMSBoxesBatchedImpl`.

> Also, why are max_coord and offset floats in case of Rect_<double>?

Typos, they should be double. Fixed in latest commit.